mantidproject/mantid
# Improvements to DiscusCalculateMultipleScatteringCorrection to support iterative process

**Description of work.**

Improvements to DiscusCalculateMultipleScatteringCorrection to support iterative process:
- performance improvements, especially if importance sampling is on:
  - the QS(Q) profile is prepared once per spectrum rather than once per bin
  - a histogram object called invPOfQ is now created once per spectrum and reused in each pass through a bin for loop rather than being created\destroyed on each pass through
- One small functional change to DiscusCalculateMultipleScatteringCorrection:
  - approach for converting S(Q) supplied as histogram to a points dataset has been modified to be consistent with flat interpolation used in algorithm
- extended the workspace creation helper:
  - expose some more parameters that control the bin sizes rather than hardcoding them (x0, deltax)
  - added an extra level to the "geographical" instrument definition (a ObjCompAssembly) so that the instrument can be saved to a Nexus file using SaveNexusESS. This means I can save instruments created by the helper without having to have an IDF file

Note the iterative process itself isn't part of this - that's a more "experimental" python script that I've not pushed yet

**To test:**

Run the following script. On the new code it completes in about 5 seconds for me on a Release build. It used to take about 1 min 30 secs (eg on the recently released 6.3 version of Mantid). Note - it should be quick because it uses a single scenario:

```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

test_ws = Load("PEARL00115245")
test_ws = ConvertUnits(test_ws, "Wavelength")
X=[5,6] # flat isotropic S(Q)
Y=[1,1] 
Sofq=CreateWorkspace(DataX=X,DataY=Y,UnitX="MomentumTransfer")
SetSample(InputWorkspace=test_ws,Geometry={'Shape': 'Cylinder', 'Height': 1.0,
                    'Radius': 0.2985, 'Center': [0.,0.,0.]},
          Material={'ChemicalFormula': 'Ce-O2', 'MassDensity': 7.215})
DiscusMultipleScatteringCorrection(test_ws,Sofq, ImportanceSampling=True, NeutronPathsSingle=1, NeutronPathsMultiple=1,
                                   OutputWorkspace="MSResults")
```

Fixes #33049.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
